python/cucim/src/cucim/core/operations/color/stain_normalizer.py

```diff
@@ -0,0 +1,821 @@
+    if method == 'lstsq':
+        if src_stain_coeff.shape != (3, 2):
+            raise ValueError("expected a set of two stain vectors")
+        # If there are only two stain vectors, use a least-squares estimate.
+        conc_raw = cp.linalg.lstsq(
+            src_stain_coeff, absorbance, rcond=None
+        )[0]
+    elif method == 'ortho':
+        if src_stain_coeff.shape != (3, 3):
+            raise ValueError("expected a set of three stain vectors")
+        # This approach relies on a square stain coeffs matrix as used by
+        # HistomicsTK. In practice, it gives nearly identical results to the
+        # least-squares approach.
+        coeff_inv = cp.linalg.inv(src_stain_coeff)
+        conc_raw = cp.dot(cp.asarray(coeff_inv, order='F'), absorbance)[:2]
+    else:
+        raise ValueError(f"unknown method, {method}")
+    return conc_raw
+
+
+def _normalized_from_concentrations(conc_raw, max_percentile, ref_stain_coeff,
+                                    ref_max_conc, source_intensity,
+                                    original_shape, channel_axis):
+    """Determine normalized image from concentrations."""
+
+    # verify conc_raw is shape (2, n_pixels)
+    if conc_raw.ndim != 2 or conc_raw.shape[0] != 2:
+        raise ValueError(
+            "`conc_raw` must be a 2D array of concentrations with size 2 on "
+            "axis 0."
+        )
+    if ref_stain_coeff.shape != (3, 2):
+        raise ValueError(
+            "`ref_stain_coeff` must be a shape (3, 2) matrix, representing "
+            "two stain vectors."
+        )
+    if len(ref_max_conc) != 2:
+        raise ValueError("`ref_max_conc` must have length 2.")
+
+    # normalize stain concentrations
+    # Note: calling percentile separately for each channel is faster than:
+    #       max_conc = cp.percentile(conc_raw, 100 - alpha, axis=1)
+    max_conc = cp.concatenate(
+        [cp.percentile(ch_raw, max_percentile)[np.newaxis]
+         for ch_raw in conc_raw]
+    )
+    normalization_factors = ref_max_conc / max_conc
+    conc_norm = conc_raw * normalization_factors[:, cp.newaxis]
+
+    # reconstruct the image based on the reference stain matrix
+    absorbance_norm = ref_stain_coeff.dot(conc_norm)
+    image_norm = absorbance_to_image(
+        absorbance_norm, source_intensity=source_intensity, dtype=np.uint8
+    )
+
+    # restore original shape for each channel
+    channel_shape = (
+        original_shape[:channel_axis] + original_shape[channel_axis + 1:]
+    )
+    image_norm = cp.reshape(image_norm, (3,) + channel_shape)
+
+    # move channels from axis 0 to channel_axis
+    if channel_axis != 0:
+        image_norm = cp.moveaxis(
+            image_norm, source=0, destination=channel_axis
+        )
+    # restore original shape
+    return image_norm
+
+
+def normalize_colors_macenko(
+        image,
+        source_intensity: float = 240.0,
+        alpha: float = 1.0,
+        beta: float = 0.15,
```

Comment:
As you mentioned before since we are using `log` instead of `log10`, maybe we should adjust this value to `0.15/log10(e) ~ 0.345`, what do you think?

Reply:
I will change it and add a note to indicate it is equivalent to the default in the publication